man/omorfi-freq-evals.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.46.6.
.TH OMORFI-FREQ-EVALS "1" "June 2018" "OMORFI" "User Commands"
.SH NAME
omorfi-freq-evals \- Count analysis coverage of Finnish text with omorfi
.SH SYNOPSIS
.B omorfi-freq-evals.bash
[\fI\,OPTION\/\fR] [\fI\,FILENAME\/\fR...]
.SH DESCRIPTION
Creates a frequency list of running text and counts coverage using omorfi.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print this help dialog
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-X\fR, \fB\-\-all\-words\fR
Use large\-coverage dictionary
.PP
If no FILENAMEs are given, input is read from standard input.
.SH EXAMPLE
Turn corpus into coverage statistics:
omorfi-freq-evals test.corpus
.SH COPYRIGHT
Copyright \(co 2018 Tommi A Pirinen
Licence GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# NEWS

This file lists noteworthy changes between releases, for full list of changes,
see git log and then `ChangeLog.old`.

## Significant changes in 20180511

* Universal dependencies version 2.2 is now used as target
* At least 226 new words: 239 additions and 13 deletions in lexeme database
* Most changes are in development infra, so not visible to end users...
* Started rewriting CG from the scratch
* The APIs for programming language deprecate load(filename) and load(dir) forms
  of filename guessing functions in favour of forthcoming loadAnalyser(file),
  loadLemmatiser(file), loadUDPipe(file) etc. etc. functions
* Working towards more general tokenise-analyse-disambiguate pipelines maybe,
  or just refactoring
* lots more automated tests -> lots less human errors
* By popular request: there are two analysers now, one with small dictionary
  and one with full, use the smaller one when you do not want to see birds or
  languages or tribes analysed. The smaller one replaces the old default, but
  the new tools will require you to select one explicitly anyways
* fixes and workarounds: java and c++ can now be disabled partially or totally
* adopted SG0 as possible verb form analysis from UD data
* The end users are now provided with bash-scripts wrappers for all
  functionalities, whereas the typically python versions allow more control
  of parametres


## Significant changes in 20170515

* Universal Dependencies version 2 is now used, still mainly lemma, UPOS,
  features fields are analysed
* At least 2,336 new words (based on diffstat: 38886 additions, 3655 deletions)
* Preliminary support for various guessing models: python-based, finite-state
  and UDPipe. This means that it is possibly to get analyses for all tokens,
  albeit quality of guesses varies.
* A minimal C++ library version has been made to match java and python bindings.
  C++-11 and libhfst are required.
* The dix version can now be compiled with lttoolbox with a lot of memory
* A restricted "gold" dictionary mode has been added. This is good for both end
  users with limited memory and end users who require higher quality lexemes
  (i.e., only research institute approved, no wiktionary words or other weird
   stuffs)
* Documentations and automatic testing much reworked with the new modern toys
  from github: travis-ci, jekyll
* Started weeding the ADP/ADV jungle...
* Fixed a horrible bug in the corpus coverage testing that terribly
  under-estimated our coverage for corpora where hapax legomena etc. were
  ignored
* Lot of documentation has been semi-automated, therefore many changes can be
  viewed at the new gh-pages site: https://flammie.github.io/omorfi/

## Significant changes in 20161115

* Started drafting more blacklists and *known good* lexemes subsets for people
  who struggle with rare words and productive compounding, derivation
* Updated to Universal Dependencies version 1.4
* A lot of new derivations by the way
* Preliminary guessers
* More loopy guessery things for punctuation and digit combos
* Minor fixes to UD feature sorting
* Homonym numbers used in some applications
* Added timeouts where downstream tools support them, so tools don't seem like
  they are freezing at random
* moved old documentations to github-pages
* added preliminary hfst-pmatch-based tokeniser

## Significant changes in 20160515

* Universal Dependencies for Finnish is the new standard format we now follow:
    * POS is now UPOS and classes were changed accordingly (new classes: AUX,
      PROPN, DET, CONJ, SCONJ, PUNCT, SYM, and VERB, NOUN, ADP, ADV as before)
    * other features mostly match the feature field in UD documentation
    * release cycle aims to be same six month cycle as with UD
    * the automatic tests verify compatibility with UD; 92 % of lemmas, primary
      POS tags and morphological features are the same as Finnish UD corpus,
      75 % same as Finnish FTB UD corpus
    * analyser for reading and writing CONLL-U format
* tokenisation as script and more hacks to token stripping in corner cases
* continuous integration with travis-ci, currently only testing basic script
  programming conventions
* added a lot of high coverage words and forms by hand
* by popular request, some of the words can now be blacklisted, when you don't
  want that guy named Mutta to ambiguate your conjunction analyses or the odd
  new guinean bird to clash with some common verb
* the "database" is now only keyed on lemma + homonym number; paradigm is extra
  information like anything else
* a lot of work on morphological segmentation towards statistical machine
  translation; check proceedings of WMT shared tasks 2015 and 2016 to see why
* started refactoring some python code into classes

## Significant changes in 20150904

* allomorphy can be tagged again to distinguish e.g. *-iden* and *-itten* when
  generating
* FinnTreeBank-1 format provided by Miikka Silfverberg is available but not
  built by default since it lacks a test set
* lexicalised inflections can have separate tag, e.g. *kännissä* can be lexical
  inessive distinguished from regular inessive
* preliminary VISL CG-3 support, with original grammar by Fred Karlsson;
  convenience bash scripts available for disambiguated parsing
* preliminary support for conllu and conllx analysis formats
* paradigm categorisation is now verified by regular expressions
* lots of paradigm fixes and some added words

## Significant changes in 20150326

* speed is up to >20,000 tokens per second from ~500
* coverages are up to:
  europarl (99 %) gutenberg (97 %), JRC Acquis (94 %) and fiwiki (93 %)
* moses factored model format supported
* segmentation supported
* Java API
* Python hacks packaged to API and module
* Rest of hand-written Xerox legacy data removed; all is script-generated
* github migration since google code is EOL'd
* file naming for automata changed to include omorfi prefix for all file
  names in case they are distributed separately.

## Significant changes in 20141014


src/python/omorfi-freq-evals.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Test sort | uniq -c | sort -nr'd gold analysis data for faithfulness
"""


from argparse import ArgumentParser, FileType
from sys import stderr, stdout, stdin
from time import perf_counter, process_time

from omorfi.omorfi import Omorfi
from omorfi.token import is_tokenlist_oov, format_feats_ftb, get_lemmas

def main():
    a = ArgumentParser()
    a.add_argument('-a', '--analyser', metavar='FSAFILE', required=True,
                   help="load analyser from FSAFILE")
    a.add_argument('-i', '--input', metavar="INFILE", type=open,
                   dest="infile", help="source of analysis data")
    a.add_argument('-o', '--output', metavar="OUTFILE",
                   type=FileType('w'),
                   dest="outfile", help="log outputs to OUTFILE")
    a.add_argument('-X', '--statistics', metavar="STATFILE",
                   type=FileType('w'),
                   dest="statfile", help="statistics")
    a.add_argument('-v', '--verbose', action="store_true", default=False,
                   help="Print verbosely while processing")
    a.add_argument('-C', '--no-casing', action="store_true", default=False,
                   help="Do not try to recase input and output when matching")
    a.add_argument('-f', '--format', metavar="FORMAT",
                   help="use FORMAT formatter to compare analyses",
                   choices=["coverage", "ftb3.1"], default="coverage")
    a.add_argument('-c', '--count', metavar="FREQ", default=0,
                   help="test only word-forms with frequency higher than FREQ")
    a.add_argument('-t', '--threshold', metavar="THOLD", default=99,
                   help="if coverage is less than THOLD exit with error")
    options = a.parse_args()
    omorfi = Omorfi(options.verbose)
    try:
        if options.analyser:
            if options.verbose:
                print("reading analyser from", options.analyser)
            omorfi.load_analyser(options.analyser)
        if not options.infile:
            options.infile = stdin
            print("reading from <stdin>")
        if not options.statfile:
            options.statfile = stdout
        if not options.outfile:
            options.outfile = stdout
    except IOError:
        print("Could not process file", options.analyser, file=stderr)
        exit(2)
    # basic statistics
    covered = 0
    full_matches = 0
    lemma_matches = 0
    anal_matches = 0
    no_matches = 0
    no_results = 0
    lines = 0
    # types
    types_covered = 0
    types_no_results = 0
    types = 0
    # for make check target
    threshold = options.threshold
    realstart = perf_counter()
    cpustart = process_time()
    for line in options.infile:
        fields = line.strip().replace(' ', '\t', 1).split('\t')
        if len(fields) < 2:
            print("ERROR: Skipping line", fields, file=stderr)
            continue
        freq = int(fields[0])
        if freq < int(options.count):
            break
        surf = fields[1]
        lemma = surf
        analysis = surf
        if options.format != 'coverage':
            lemma = fields[2]
            analysis = fields[3]
        lines += freq
        types += 1
        if options.verbose:
            print(lines, '(', freq, ') ...', end='\r')
        anals = omorfi.analyse(surf)
        if not is_tokenlist_oov(anals):
            covered += freq
            types_covered += 1
        else:
            no_results += freq
            types_no_results += 1
            print(freq, "OOV", surf, sep='\t', file=options.outfile)
        found_anals = False
        found_lemma = False
        for anal in anals:
            if options.format == 'ftb3.1':
                anal_ftb3 = format_feats_ftb(anal)
                lemma_ftb3 = '#'.join(get_lemmas(anal))
                # hacks ftb3:
                analysis = analysis.replace(" >>>", "")
                if analysis == anal_ftb3:
                    found_anals = True
                elif set(anal_ftb3.split()) == set(analysis.split()):
                    found_anals = True
                    print(freq, "PERMUTAHIT", analysis, anal_ftb3, sep='\t',
                            file=options.outfile)
                else:
                    print(freq, "ANALMISS", analysis, anal_ftb3, sep='\t',
                            file=options.outfile)
                if lemma == lemma_ftb3:
                    found_lemma = True
                elif lemma.replace('#', '') == lemma_ftb3.replace('#', ''):
                    found_lemma = True
                    print(freq, "LEMMARECOMP", lemma, lemma_ftb3, sep='\t',
                            file=options.outfile)
                else:
                    print("LEMMAMISS", lemma, lemma_ftb3, sep='\t',
                            file=options.outfile)
        if options.format != 'coverage':
            if not found_anals and not found_lemma:
                no_matches += freq
                print(freq, "NOHITS!", surf, sep='\t', file=options.outfile)
            elif found_anals and found_lemma:
                full_matches += freq
            elif not found_anals:
                anal_matches += freq
                print(freq, "LEMMANOANAL", surf, sep='\t', file=options.outfile)
            elif not found_lemma:
                lemma_matches += freq
                print(freq, "ANALNOLEMMA", surf, sep='\t', file=options.outfile)
            else:
                print("Logical error, kill everyone")
                exit(13)
    realend = perf_counter()
    cpuend = process_time()
    print("CPU time:", cpuend - cpustart, "real time:", realend - realstart)
    print("Lines", "Covered", "OOV", sep="\t", file=options.statfile)
    print(lines, covered, lines-covered, sep="\t", file=options.statfile)
    print(lines / lines * 100 if lines != 0 else 0,
          covered / lines * 100 if lines != 0 else 0,
          (lines-covered) / lines * 100 if lines != 0 else 0,
          sep="\t", file=options.statfile)
    print("Types", "Covered", "OOV", sep="\t", file=options.statfile)
    print(types, types_covered, types-types_covered, sep="\t", file=options.statfile)
    print(types / types * 100 if types != 0 else 0,
          types_covered / types * 100 if types != 0 else 0,
          (types-types_covered) / types * 100 if types != 0 else 0,
          sep="\t", file=options.statfile)
    if options.format == 'ftb3.1':
        print("Lines", "Matches", "Lemma", "Anals", "Mismatch", "No results", sep="\t",
              file=options.statfile)
        print(lines, full_matches, lemma_matches, anal_matches, no_matches,
              no_results,
              sep="\t", file=options.statfile)
        print(lines / lines * 100 if lines != 0 else 0,
              full_matches / lines * 100 if lines != 0 else 0,
              lemma_matches / lines * 100 if lines != 0 else 0,
              anal_matches / lines * 100 if lines != 0 else 0,
              no_matches / lines * 100 if lines != 0 else 0,
              no_results / lines * 100 if lines != 0 else 0,
              sep="\t", file=options.statfile)
    if lines == 0:
        print("Needs more than 0 lines to determine something",
              file=stderr)
        exit(2)
    elif options.format == 'ftb3.1' and \
            (full_matches / lines * 100 <= int(options.threshold)):
        print("needs to have", threshold, "% matches to pass regress test\n",
              "please examine", options.outfile.name, "for regressions",
              file=stderr)
        exit(1)
    elif options.format == 'coverage' and \
            (covered / lines * 100 <= int(options.threshold)):
        print("needs to have", threshold, "% coverage to pass regress test\n",
              "please examine", options.outfile.name, "for regressions",
              file=stderr)
        exit(1)
    else:
        exit(0)


if __name__ == "__main__":
    main()
